core/app/api/session-sync/[jwt]/route.ts

```diff
@@ -0,0 +1,117 @@
+    validateSessionSyncJwt(jwt: $jwt) {
+      content {
+        redirectTo
+        customer {
+          entityId
+          firstName
+          lastName
+          email
+        }
+        customerAccessToken {
+          value
+          expiresAt
+        }
+        cart {
+          entityId
+        }
+      }
+      errors {
+        __typename
+        ... on InvalidSessionSyncJwtError {
+          errorType
+          message
+        }
+        ... on JwtTokenExpiredError {
+          message
+        }
+      }
+    }
+  }
+`);
+
+export const GET = async (
+  request: NextRequest,
+  { params }: { params: Promise<{ jwt: string }> },
+) => {
+  const { jwt } = await params;
+  const query = request.nextUrl.searchParams;
+
+  const { data, errors } = await client.fetch({
+    document: VerifySessionSyncJwtMutation,
+    variables: { jwt },
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  // Handle any API Errors
+  if (errors?.length) {
+    return new NextResponse('', { status: 500, statusText: 'Server error' });
+  }
+
+  const { content, errors: validationErrors } = data.validateSessionSyncJwt;
+
+  // Handle any validation errors or if the content is null
+  if (validationErrors.length || content === null) {
+    const error = validationErrors.at(0);
+
+    switch (error?.__typename) {
+      case 'InvalidSessionSyncJwtError':
+        return new NextResponse(null, { status: 400, statusText: 'Bad request' });
+
+      case 'JwtTokenExpiredError':
+        return NextResponse.redirect(request.referrer);
+
+      default:
+        return new NextResponse(null, { status: 500, statusText: 'Server error' });
+    }
+  }
+
+  // Update the cart id if it exists regardless of if the customer is logged in or not
+  if (content.cart?.entityId) {
+    await setCartId(content.cart.entityId);
+  } else {
+    await clearCartId();
+  }
+
+  // Soon there will always be a session, whether guest or logged in.
+  // We will need to remove this check once that work is complete. We can just
+  // update the session in that case.
+  if (content.customer && content.customerAccessToken) {
+    await signIn('session-sync', {
+      name: `${content.customer.firstName} ${content.customer.lastName}`,
+      email: content.customer.email,
+      customerAccessToken: content.customerAccessToken.value,
+      redirect: false,
+    });
+  } else {
+    await signOut({ redirect: false });
+  }
+
+  // We want to revalidate all the data as we are either logged in or out.
+  revalidatePath('/', 'layout');
+
+  if (query.get('redirect') === 'false') {
+    return new NextResponse(null, {
+      status: 204,
+      headers: {
+        'Access-Control-Allow-Origin': 'https://session-sync-example.vercel.app',
+        'Access-Control-Allow-Credentials': 'true',
+      },
+    });
+  }
+
+  // Note: redirectTo is going to include the full url, not a partial path
+  return NextResponse.redirect(content.redirectTo);
```

Comment:
We need to validate `redirectTo`, making sure it is an expected value. Otherwise, we would be vulnerable to [open redirect attack](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html#dangerous-url-redirects) (e.g.: phishing). We could validate the host to make sure it matches the host of BC checkout of the same channel.

Reply:
I'll add some validation here – I did some testing with some bad urls and I think we have some protection already as BigCommerce prepends the merchant-provided url with either the storefront or checkout domain. The only use-case I would be concerned about is if a merchant has the domain `https://foo.co` and bad actor has `https://foo.com`, they can set the `redirectTo` field to `m/bad-path` and it'll be redirected to `https://foo.com/bad-path`.

Reply:
Thanks for checking, much appreciated! There's also the possibility that an attacker could construct the URL themselves (instead of BC) and trick users into navigating to it as part of a phishing attack. Therefore, even if we have existing protections on the BC side for storing the checkout domain, it's still important to have validation in place for the session-sync endpoint on the Catalyst side.